File: implementations/opshin/src/nft_minting_policy.py
#!opshin
from opshin.prelude import *


def int_to_bs(n: int) -> List[int]:
    acc: List[int] = []
    while n != 0:
        if n < 256:
            acc = acc + [n]
            n = 0
        else:
            acc = acc + [n % 256]
            n = n // 256
    return acc


def derive_nft_name(ref: TxOutRef) -> TokenName:
    return ref.id.tx_id + bytes(int_to_bs(ref.idx))


def validator(initial_spend: TxOutRef, context: ScriptContext) -> None:
    purpose = context.purpose
    if isinstance(purpose, Minting):
        own_symbol = purpose.policy_id
    else:
        assert False, "Invalid purpose"

    minted_names = context.tx_info.mint[own_symbol].items()
    assert len(minted_names) == 1, "Minted ore than one NFT name"
    minted_nft_name, minted_nft_amount = minted_names[0]

    assert minted_nft_amount == 1, "Minted one NFT"
    assert minted_nft_name == derive_nft_name(
        initial_spend
    ), "Minted NFT name does not match derived name"

    assert any(
        [initial_spend == inp.out_ref for inp in context.tx_info.inputs]
    ), "Initial spent is not spent"

File: implementations/plutus-tx/lib/UplcBenchmark/NftMintingPolicy.hs
{-# LANGUAGE NoImplicitPrelude #-}
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}

module UplcBenchmark.NftMintingPolicy (nftMintingPolicy) where

import Data.Kind (Type)
import PlutusLedgerApi.V3 (
  BuiltinByteString,
  BuiltinData,
  FromData (fromBuiltinData),
  ScriptContext (scriptContextRedeemer, scriptContextScriptInfo, scriptContextTxInfo),
  ScriptInfo (MintingScript),
  TokenName (TokenName),
  TxId (getTxId),
  TxInInfo (txInInfoOutRef),
  TxInfo (txInfoInputs, txInfoMint),
  TxOutRef (txOutRefIdx),
  Value (Value),
  getRedeemer,
  txOutRefId,
  unsafeFromBuiltinData,
 )
import PlutusLedgerApi.V3.MintValue (MintValue (UnsafeMintValue))
import PlutusTx.Prelude (
  BuiltinUnit,
  Integer,
  any,
  appendByteString,
  check,
  consByteString,
  divide,
  emptyByteString,
  modulo,
  traceIfFalse,
  ($),
  (&&),
  (<),
  (==),
 )
import UplcBenchmark.Utils (fromJustTrace, getOwnMint)

type NftMintingPolicyRedeemer :: Type
newtype NftMintingPolicyRedeemer = NftMintingPolicyRedeemer'CreatePool TxOutRef
  deriving newtype (FromData)

{-# INLINE integerToByteString #-}
integerToByteString :: Integer -> BuiltinByteString
integerToByteString = go emptyByteString
  where
    {-# INLINE go #-}
    go rest n =
      if n < 256
        then consByteString n rest
        else go (consByteString (n `modulo` 256) rest) (n `divide` 256)

{-# INLINE deriveNftName #-}
deriveNftName :: TxOutRef -> TokenName
deriveNftName txOutRef =
  let idPart = getTxId $ txOutRefId txOutRef
      idxPart = integerToByteString $ txOutRefIdx txOutRef
   in TokenName (appendByteString idPart idxPart)

{-# INLINE nftMintingPolicy #-}
nftMintingPolicy :: BuiltinData -> BuiltinUnit
nftMintingPolicy rawCtx =
  let !ctx :: ScriptContext = unsafeFromBuiltinData rawCtx
      !redeemer :: NftMintingPolicyRedeemer =
        fromJustTrace "Could not decode redeemer" $ fromBuiltinData $ getRedeemer $ scriptContextRedeemer ctx

      NftMintingPolicyRedeemer'CreatePool !initialSpend = redeemer

      !expectedTokenName = deriveNftName initialSpend

      !txInfo = scriptContextTxInfo ctx
      !inputs = txInfoInputs txInfo
      UnsafeMintValue mint = txInfoMint txInfo

      MintingScript !ownSymbol = scriptContextScriptInfo ctx
      ![(!mintedNftName, !mintedNftAmount)] = getOwnMint ownSymbol (Value mint)

      isInitialSpend txInInfo = txInInfoOutRef txInInfo == initialSpend

      !mintingOneNft = traceIfFalse "Minted more than one NFT" (mintedNftAmount == 1)
      !mintingNftNameCorrect =
        traceIfFalse
          "Minted NFT name doesn't match initial spend"
          (mintedNftName == expectedTokenName)
      !spendingInitialSpend =
        traceIfFalse
          "Initial spend was not spent"
          (any isInitialSpend inputs)
   in check (mintingOneNft && mintingNftNameCorrect && spendingInitialSpend)
